src/components/utils/test/thread_test.cc

```diff
@@ -175,6 +247,175 @@ TEST_F(ThreadTest, StartOneThreadTwice_ExpectTheSameThreadStartedTwice) {
+  delegate->notified_ = false;
+  delegate->notify_running_ = false;
+  EXPECT_TRUE(thread->start());
+
+  // Wait for the thread to notify it is running
+  {
+    AutoLock test_lock(test_mutex_);
+    while (!delegate->notify_running_) {
+      cond_var_.Wait(test_lock);
+    }
+  }
+  EXPECT_TRUE(thread->is_running());
+
+  // Make the thread stop and wait for it
+  delegate->notified_ = true;
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  delete delegate;
+  DeleteThread(thread);
+}
+
+TEST(ThreadTest2, StartJoinStart_Success) {
+  // Create a delegate with infinite threadMain
+  InfiniteMainTestDelegate* delegate = new InfiniteMainTestDelegate();
+  Thread* thread = CreateThread(kThreadName.c_str(), delegate);
+  EXPECT_TRUE(thread->start(threads::ThreadOptions()));
+  EXPECT_TRUE(thread->is_joinable());
+
+  // Wait for the thread to notify it has started execution
+  {
+    AutoLock test_lock(test_mutex_);
+    while (!delegate->notify_running_) {
+      cond_var_.Wait(test_lock);
+    }
+  }
+  EXPECT_TRUE(thread->is_running());
+  // Signal the delegate it has to finish the execution of threadMain
+  delegate->notified_ = true;
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  // Reset the parameters of the delegate so thread can be started again
+  delegate->notified_ = false;
+  delegate->notify_running_ = false;
+
+  EXPECT_TRUE(thread->start());
+
+  // Wait for the thread to notify it has started execution
+  {
+    AutoLock test_lock(test_mutex_);
+    while (!delegate->notify_running_) {
+      cond_var_.Wait(test_lock);
+    }
+  }
+  EXPECT_TRUE(thread->is_running());
+
+  // Signal the delegate it has to finish the execution of threadMain
+  delegate->notified_ = true;
+  thread->join(Thread::JoinOptionStop::kNoStop);
+  EXPECT_FALSE(thread->is_running());
+
+  delete delegate;
+  DeleteThread(thread);
+}
+
+TEST(ThreadTest2, DeleteJoinable_Success) {
+  // Create a delegate with delayed exit of threadMain
+  DelayedExitTestDelegate* delegate = new DelayedExitTestDelegate();
+  Thread* thread = CreateThread(kThreadName.c_str(), delegate);
+  EXPECT_TRUE(thread->start(threads::ThreadOptions()));
+  EXPECT_TRUE(thread->is_joinable());
+
+  // Wait for the thread to notify it has started execution
+  {
+    AutoLock test_lock(test_mutex_);
+    while (!delegate->notify_running_) {
+      cond_var_.Wait(test_lock);
+    }
+  }
+
+  // Reset the parameters of the delegate
+  delegate->notify_running_ = false;
+  EXPECT_TRUE(thread->is_running());
+  DeleteThread(thread);
+
+  // Wait for threadMain to notify it is exiting
+  {
+    AutoLock test_lock(test_mutex_);
+    while (!delegate->notify_running_) {
+      cond_var_.Wait(test_lock);
+    }
+  }
+  EXPECT_TRUE(delegate->notified_exit_);
+
+  // Creating a new thread so the delegate can be deleted
+  Thread* thread2 = CreateThread(kThreadName.c_str(), delegate);
```

Comment:
@dev-gh In `ThreadDelegate` destructor there is a call to the `Thread` object which was deleted above.